Initial FreeBSD support

pescador-barquero opened the pull request:
Editor and samples work.

There are a few little changes required to build on FreeBSD. 

* sysinfo is provided by http://freshports.org/devel/libsysinfo
* pthread*_np functions are found in <pthread_np.h>
* CMake stores system processor name as amd64. Which other systems report it as AMD64?
* Clang fails to find the OpenImageDenoise library as instructed in pragmas. But it builds and links against OIDN with those pragmas commented.
* Provided libdxcompiler.so is made for Linux systems so I recompiled it from [Github](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/BuildingAndTestingDXC.rst) and replaced the file. I have made no modification in the build system with this regard.

brakhane commented on WickedEngine/wiJobSystem.cpp:
```diff
@@ -14,6 +14,9 @@
 
 #ifdef PLATFORM_LINUX
 #include <pthread.h>
+#if __has_include(<pthread_np.h>)
```
pthread_np seems FreeBSD exclusive, so wouldn't it make more sense to use `#if __FREEBSD__` here? Since if we're under BSD, that header is needed for compilation, and not just optional.

brakhane commented on WickedEngine/wiJobSystem.cpp:
```diff
@@ -198,7 +201,8 @@ namespace wi::jobsystem
 
-#ifdef PLATFORM_LINUX
+#if defined(__FREEBSD__)
+#elif defined(PLATFORM_LINUX)
```
A short `//TODO` comment wouldn't hurt

brakhane reviewed (commented):
I'm thinking about if we should add a PLATFORM_FREEBSD macro that implies PLATFORM_LINUX (CMake would set both) or keep the current "`PLATFORM_LINUX`+`__FREEBSD__` mix. Both feel kinda ugly.

Maybe having `PLATFORM_FREEBSD` and changing `#if PLATFORM_LINUX` to `#if defined(PLATFORM_LINUX) || defined(PLATFORM_FREEBSD)` would be the cleanest way.

This is just me brainstorming, not a "please do it" comment. What do you think?

brakhane commented on WickedEngine/CMakeLists.txt:
```diff
@@ -139,7 +144,7 @@ if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "MSVC")
         endif()
     endif()
-elseif ("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "AMD64" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "i386")
+elseif ("${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86_64" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "AMD64" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "amd64" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "x86" OR "${CMAKE_SYSTEM_PROCESSOR}" STREQUAL "i386")
```
I mean, if it doesn't work correctly on Wicked, it will not work correctly in Jolt either. It seems like an oversight in Jolt, so would be a bugfix

brakhane commented:
> Provided libdxcompiler.so is made for Linux systems so I recompiled it from [Github](https://github.com/microsoft/DirectXShaderCompiler/blob/main/docs/BuildingAndTestingDXC.rst) and replaced the file. I have made no modification in the build system with this regard.

@pescador-barquero that file should then be named something like `libdxcompiler-freebsd.so` or so and CMake could be modified to then copy it as libdxcompiler.so.

Did you compile from the official release, or current master? We use the official release version for Wicked.

It would also make sense to add the current version to the filename, because when a new DXC version comes out, we most likely will not be updating the freebsd version alongside it; having the version number easily visible would make it clear that it needs to be updated later on.

pescador-barquero commented:
* Removed the use of libsysinfo 
* Removed the use of pkg-config

I made an attempt at handling the libdxcompiler situation for FreeBSD. Without a deep CMakeLists rework it ends up as awkward special case code. I also tried to get the Wayland PR a test but I ran into build system problems after solving C++ git rebase conflicts.

I would like to defer libdxcompiler blob CMake support for another time or developer.

turanszkij commented on WickedEngine/wiRenderPath3D_PathTracing.cpp:
```diff
@@ -12,7 +12,7 @@
 #if OIDN_VERSION_MAJOR >= 2
 #define OPEN_IMAGE_DENOISE
-#pragma comment(lib,"OpenImageDenoise.lib")
+//#pragma comment(lib,"OpenImageDenoise.lib")
```
Sorry I missed this, yes it's fine to have windows only pragma comment for openimagedenoise

brakhane reviewed (approved):
As already mentioned, I would prefer to raise an exception instead of returning an empty string in case the exe path lookup fails, but if the current version is ok for János I don't insist on it. Rest of it looks good.

setpriority handling is still a noop, so worker threads will run with the same priority, but that can be done later (IIRC Linux version also didn't have it for a long time).
